python/ray/experimental/data/impl/arrow_block.py

```diff
@@ -89,6 +99,13 @@ def add(self, item: Union[dict, ArrowRow]) -> None:
         for key, value in item.items():
             self._columns[key].append(value)
 
+    def add_pandas_df(self, item: "pandas.DataFrame"):
+        item = pyarrow.Table.from_pandas(item).to_pydict()
+        # Pandas has a column: List[Row] structure.
+        for col, rows in item.items():
+            for row in rows:
+                self._columns[col].append(row)
```

Comment:
Change this to use `.add_block()` instead (I have a separate fix that makes add_block much more efficient than adding individual rows).